tensorflow/core/kernels/cast_op.h

```diff
@@ -43,6 +43,111 @@ struct CastFunctor {
 namespace Eigen {
 namespace internal {
 
+// Eigen can't convert to/from complex numbers, because it is limited to cases
+// that can be static_casted. But numpy is able to cast to/from complex, which
+// we want to emulate. So we add specializations for complex here.
+typedef std::complex<float> complex64;
+typedef std::complex<double> complex128;
+using tensorflow::uint8;
+using tensorflow::int8;
+using tensorflow::uint16;
+using tensorflow::int16;
+using tensorflow::int32;
+using tensorflow::int64;
+
+// These are seperate definitions of what should happen when we cast from/to
+// complex numbers. The actual template specializations are instantiated below.
+template<typename From, typename To>
+struct cast_from_complex_impl {
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE To operator()(const From& a) const {
+    return static_cast<To>(a.real());
+  }
+};
+
+template<typename From, typename To>
```

Comment:
The code would be simpler if you used partial specialization instead: you could for example write this to implement the casting from complex to real.
template <typename From, typename To>
struct scalar_cast_op<std::complex<From>, To> {
 EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE To operator()(const std::complex<From>& a) const {
  return static_cast<To>(a.real());
  }
};